### PR TITLE
make sure HEAD requests do not return a body

### PR DESCRIPTION
**Pull request by lordrhodos**
Suggests a fix for issue #2386 

@akrabat  @l0gicgate I need some feedback if this is the right approach

### CODE REVIEW

File: Slim/App.php [hunk 1]
```diff
@@ -645,6 +647,11 @@ protected function finalize(ResponseInterface $response)
      */
     protected function isEmptyResponse(ResponseInterface $response)
     {
+        $request = $this->container->get('request');
+        if ($request->getMethod() === 'HEAD') {
```
Comment (l0gicgate) [inline Slim/App.php]: You will need to make this if case insensitive. This will fail with lower case method 'head'.

File: Slim/App.php [hunk 2]
```diff
@@ -614,7 +614,8 @@ protected function finalize(ResponseInterface $response)
         // stop PHP sending a Content-Type automatically
         ini_set('default_mimetype', '');
 
-        if ($this->isEmptyResponse($response)) {
+        $request = $this->container->get('request');
+        if ($this->isEmptyResponse($response) && $request->getMethod() !== 'HEAD') {
```
Comment (l0gicgate) [inline Slim/App.php]: this also needs to be a case insensitive if case.

**Comment by l0gicgate** [comment]
A couple of things:

1. You will need to clear out the existing Response body in `App::finalize()`;
2. Add the test case from the `4.x` branch as seen [here](https://github.com/slimphp/Slim/blob/4.x/tests/AppTest.php#L1454-L1470)

Thank you!

**Review by l0gicgate — APPROVED** [review]
Everything looks good to me @lordrhodos. Just need @akrabat to have a glance and we can merge and close #2386.

**Comment by akrabat** [comment]
Im a little twitchy about using the container's Request object, but it seems the best choice here

**Review by akrabat — APPROVED** [review]

**Comment by akrabat** [comment]
Do we need this for Slim 4 too?

**Comment by l0gicgate** [comment]
@akrabat I already did this for Slim 4 see #2547

**Pull request merged**